src/ray/gcs/gcs_server/gcs_autoscaler_state_manager.cc

```diff
@@ -0,0 +1,125 @@
+// Copyright 2022 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ray/gcs/gcs_server/gcs_autoscaler_state_manager.h"
+
+namespace ray {
+namespace gcs {
+
+GcsAutoscalerStateManager::GcsAutoscalerStateManager(
+    ClusterResourceManager &cluster_resource_manager,
+    GcsResourceManager &gcs_resource_manager,
+    GcsNodeManager &gcs_node_manager)
+    : cluster_resource_manager_(cluster_resource_manager),
+      gcs_node_manager_(gcs_node_manager),
+      gcs_resource_manager_(gcs_resource_manager),
+      cluster_resource_state_version_(0),
+      last_seen_autoscaler_state_version_(0) {}
+
+void GcsAutoscalerStateManager::HandleGetClusterResourceState(
+    rpc::autoscaler::GetClusterResourceStateRequest request,
+    rpc::autoscaler::GetClusterResourceStateReply *reply,
+    rpc::SendReplyCallback send_reply_callback) {
+  RAY_CHECK(request.last_seen_cluster_resource_state_version() <=
+            cluster_resource_state_version_);
+  reply->set_last_seen_autoscaler_state_version(last_seen_autoscaler_state_version_);
+  reply->set_cluster_resource_state_version(GetNextClusterResourceStateVersion());
+
+  GetNodeStates(reply);
+  GetPendingResourceRequests(reply);
+  GetPendingGangResourceRequests(reply);
+  GetClusterResourceConstraints(reply);
+
+  // We are not using GCS_RPC_SEND_REPLY like other GCS managers to avoid the client
+  // having to parse the gcs status code embedded.
+  send_reply_callback(ray::Status::OK(), nullptr, nullptr);
```

Comment:
I think for GCS_RPC_SEND_REPLY we will need to have a `GcsStatus` in the reply protobuf message, which isn't a big deal, but would require the client to read that instead of the callback status. 

Sang suggested not to enforce that on the client.